### ISSUE TITLE
carray_from_* unused-function warnings with 0.26b0

### ISSUE THREAD

**Issue by totaam** [cython/cython#1757]
Those weren't present before.
We're building with -Wall because until now the generated code was clean.
Here's an example of failure:

```
building 'xpra.codecs.csc_swscale.colorspace_converter' extension
creating build/temp.linux-x86_64-2.7/xpra/codecs/csc_swscale
gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -I/usr/include/xpra -I/usr/include/python2.7 -c xpra/codecs/csc_swscale/colorspace_converter.c -o build/temp.linux-x86_64-2.7/xpra/codecs/csc_swscale/colorspace_converter.o -Wl,-rpath,/usr/lib64/xpra -Wall -Werror -fPIC
xpra/codecs/csc_swscale/colorspace_converter.c:12227:12: error: ‘__Pyx_carray_from_py_unsigned_long’ defined but not used [-Werror=unused-function]
 static int __Pyx_carray_from_py_unsigned_long(PyObject *__pyx_v_o, unsigned long *__pyx_v_v, Py_ssize_t __pyx_v_length) {
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
xpra/codecs/csc_swscale/colorspace_converter.c:11690:12: error: ‘__Pyx_carray_from_py_int’ defined but not used [-Werror=unused-function]
 static int __Pyx_carray_from_py_int(PyObject *__pyx_v_o, int *__pyx_v_v, Py_ssize_t __pyx_v_length) {
            ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
error: command 'gcc' failed with exit status 1
```

And:
```
building 'xpra.codecs.vpx.encoder' extension
creating build/temp.linux-x86_64-2.7/xpra/codecs/vpx
gcc -pthread -fno-strict-aliasing -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -D_GNU_SOURCE -fPIC -fwrapv -fPIC -I/usr/include/xpra -I/usr/include/python2.7 -c xpra/codecs/vpx/encoder.c -o build/temp.linux-x86_64-2.7/xpra/codecs/vpx/encoder.o -Wall -Werror -fPIC
xpra/codecs/vpx/encoder.c:21241:30: error: ‘__pyx_convert__to_py_vpx_codec_enc_cfg_t’ defined but not used [-Werror=unused-function]
             static PyObject* __pyx_convert__to_py_vpx_codec_enc_cfg_t(vpx_codec_enc_cfg_t s) {
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
xpra/codecs/vpx/encoder.c:13358:28: error: ‘__pyx_convert__from_py_vpx_codec_enc_cfg_t’ defined but not used [-Werror=unused-function]
 static vpx_codec_enc_cfg_t __pyx_convert__from_py_vpx_codec_enc_cfg_t(PyObject *__pyx_v_obj) {
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
error: command 'gcc' failed with exit status 1
```

Not sure why those functions are generated at all.
This problem only affects these two modules out of a well over a dozen.
How can I workaround this and get Cython not to generate those functions?

**Comment by robertwb** [comment]
We shouldn't be producing these if they're not used. Can you give us a minimal repro and I'll see if I can get it fixed before the release?

**Comment by totaam** [comment]
Minimal example (no idea why removing either cdef line from the object fixes things!):
```
cdef extern from "vpx/vpx_codec.h":
    ctypedef struct vpx_codec_ctx_t:
        pass

cdef extern from "vpx/vpx_encoder.h":
    ctypedef struct vpx_codec_enc_cfg_t:
        unsigned int g_usage

cdef class Encoder:
    cdef vpx_codec_ctx_t *context
    cdef vpx_codec_enc_cfg_t cfg
```
Build it with the libvpx headers installed (on Fedora 26, that's "libvpx-devel"):
```
from distutils.core import setup
from distutils.extension import Extension
from Cython.Build import cythonize

vpx_pkgconfig = {
    'libraries':            ['vpx', 'm'],
    'extra_compile_args':   ['-Wall', '-Werror'],
    }
module = Extension("vpx", ["vpx.pyx"], **vpx_pkgconfig)
setup(ext_modules = cythonize([module]))
```
Result is exactly as above.

**Comment by scoder** [comment]
I guess it's the automatic pickling support that we have now. It seems to generate the struct field conversion functions, before it decides that it cannot actually pickle the class because it contains pointers. Removing the pointer field makes it picklable and thus uses the functions, and removing the struct field enables the decision without generating the functions.

It's a bug that you can't really work around currently.

**Comment by totaam** [comment]
Is there a way to turn off pickling?
AFAIK, I don't need it for anything, so the extra pickling code would be just bloat for me anyway?

If I was desperate (which I am not), I guess I could use `uintptr_t` instead of the struct pointer in the object, and cast back and forth.

**Comment by robertwb** [comment]
This can be disabled with the auto_pickle directive, e.g. `#cython: auto_pickle=False` at the top of the file. However, we should be able to test whether conversion is possible without actually generating them, so this bug still should be fixed.

**Comment by robertwb** [comment]
This should be fixed by 29067ceba6b1370cfd9a9b85a291957c0aaa34a1
